rfc-134-govuk-wide-session-cookie-and-login.md

```diff
@@ -0,0 +1,337 @@
+  if (req.http.Cookie:govuk_session ~ "^([a-zA-Z0-9\-_]+)?\.([a-zA-Z0-9\-_]+)?\.([a-zA-Z0-9\-_]+)?$") {
+    set req.http.GOVUK-Session-ID = digest.base64url_nopad_decode(re.group.1);
+    set req.http.GOVUK-Session-Expires = digest.base64url_nopad_decode(re.group.2);
+    set req.http.GOVUK-Session-Signature = digest.base64url_nopad_decode(re.group.3);
+    set req.http.GOVUK-Session-Expected-Signature = digest.hmac_sha256(table.lookup(env, "session_key"), req.http.GOVUK-Session-ID + "." + req.http.GOVUK-Session-Expires);
+
+    if(!digest.secure_is_equal(req.http.GOVUK-Session-Signature, req.http.GOVUK-Session-Expected-Signature) || time.is_after(now, std.integer2time(std.atoi(req.http.GOVUK-Session-Expires)))) {
+      set req.http.GOVUK-End-Session = "true";
+      unset req.http.GOVUK-Session-ID;
+    }
+
+    unset req.http.GOVUK-Session-Expires;
+    unset req.http.GOVUK-Session-Signature;
+    unset req.http.GOVUK-Session-Expected-Signature;
+  } else {
+    set req.http.GOVUK-End-Session = "true";
+  }
+}
+```
+
+#### Changes to `vcl_fetch`
+
+When fetching a response from the backend, record the new
+`GOVUK-Session-ID` value, if one is returned; and end the session, if
+the user has been logged out:
+
+```vcl
+if (beresp.http.GOVUK-Session-ID) {
+  set req.http.GOVUK-Session-ID = beresp.http.GOVUK-Session-ID;
+}
+
+if (beresp.http.GOVUK-End-Session) {
+  set req.http.GOVUK-End-Session = "true";
+  unset req.http.GOVUK-Session-ID;
+}
+```
+
+If the response depends on the user session, it must either:
+
+1. Set a `Vary: GOVUK-Session-ID` header, or
+2. Set headers to prevent caching entirely
+
+#### Changes to `vcl_deliver`
+
+When delivering a response to the user, set a new cookie with a new
+expiration time, and disable shared caches outside of Fastly (both
+Fastly and the user's browser can still cache the page) if the
+response depended on the session:
+
+```vcl
+if (req.http.GOVUK-End-Session == "true") {
+    set req.http.GOVUK-Session = "";
+} else if (req.http.GOVUK-Session-ID) {
+  set req.http.GOVUK-Session-Expires = strftime({"%s"}, time.add(now, 1800s));
+  set req.http.GOVUK-Session-Signature = digest.hmac_sha256(table.lookup(env, "session_key"), req.http.GOVUK-Session-ID + "." + req.http.GOVUK-Session-Expires));
+  set req.http.GOVUK-Session = digest.base64url_nopad(req.http.GOVUK-Session-ID) + "." + digest.base64url_nopad(req.http.GOVUK-Session-Expires) + "." + digest.base64url_nopad(req.http.GOVUK-Session-Signature);
+}
+
+if (req.http.GOVUK-Session) {
+  add resp.http.Set-Cookie = "govuk_session=" + req.http.GOVUK-Session + "; secure; httponly; samesite=strict; max-age=1800; path=/";
+}
+
+if (resp.http.Vary ~ "GOVUK-Session-ID") {
+  unset resp.http.Vary:GOVUK-Session-ID;
+  set resp.http.Cache-Control:private = "";
+}
+```
+
+The `1800` here (30 minutes) is just an example.  We will work out an
+appropriate session duration with product and security input.
+
+### Create a new app to manage the auth process
```

Comment:
Yeah, I expected this to be a point of contention.

I think putting it in frontend would also work.  My reasoning for making a new app on a new machine class is that this is likely to have very different traffic to other parts of GOV.UK---lots of direct origin hits, no caching at all---so maybe we want something which is easy to scale quickly.  And given that accounts / personalisation is a big item in the spend review & roadmap it's unlikely the team would vanish and the new app fall to Platform Health.

> gem/engine

Yeah, that's another option.  I think the difficulty there is that we'd need to do without the database of session ID -> OAuth tokens.  So we'd need to store those tokens in the cookie, which suggests to me an encrypted (rather than just signed) cookie.  But we've got to do the encryption in VCL (to avoid the origin hit on every request to bump the cookie expiration), and [Fastly's VCL for crypto][1] is very digest-heavy and not so good for encryption.

[1]: https://developer.fastly.com/reference/vcl/functions/cryptographic/

Reply:
> Yeah, I expected this to be a point of contention. 

Good good :) It always seems easier to add an app than to remove one.

It strikes me that the main architectural advantage of doing this in a new app is having a shared database of attributes that multiple apps can access - is that an approach we're going to be suggest to other departments if/when they implement accounts? As I guess, this might serve as an implementation example

I did have a ponder about the app, [based off there being no user interface](https://github.com/alphagov/govuk-rfcs/pull/134/files/2d18065973d53799743a5e4ebf4f0b10ac1ca5db#diff-f5a6037308ea735f2a6e63b3080e78792499608bcad7386a49cf94799634d1be) and the [expected later need of shared code](https://github.com/alphagov/govuk-rfcs/pull/134#discussion_r567730827), and wondered if pitching it as accounts-frontend might not be ideal as it'd be the only frontend app without an interface. I wondered whether the sign-in and sign-out parts might actually belong distinct from the API aspects of the app - probably in a different frontend app. It struck me as an architectural smell having an app that is partially private API and partially public, the only app of ours I can of with this is Email Alert API and that hasn't worked out too well in keeping the private API actually private.

Reply:
> It strikes me that the main architectural advantage of doing this in a new app is having a shared database of attributes that multiple apps can access - is that an approach we're going to be suggest to other departments if/when they implement accounts?

We think there are likely to be two classes of attributes: those which are useful only within a specific service (or department) and those which are useful across government.  And those ones which aren't useful across government will live with the service.  For example, if GOV.UK Pay use GOV.UK Accounts to implement a way for users to set up and manage recurring payments, we definitely don't want to end up storing payment information in the cross-gov attribute store.

So there is a need for attributes held within a service.  And if that service is composed of microservices (like GOV.UK Publishing is), then the attributes need to be shared.

>  I wondered whether the sign-in and sign-out parts might actually belong distinct from the API aspects of the app - probably in a different frontend app. It struck me as an architectural smell having an app that is partially private API and partially public, the only app of ours I can of with this is Email Alert API and that hasn't worked out too well in keeping the private API actually private.

That's a very good point, it is a bit of a weird "frontend" app.  Maybe the login / logout functionality should live in `frontend` and the attribute management part live in a separate API service which is kept properly internal.

@erino what do you think? This ties in with your work on a GOV.UK Attribute Service.  For this RFC to work we'll need:

- A database mapping session IDs to user IDs (one user can have multiple sessions)
- A database mapping user IDs to OAuth tokens (so we can use the cross-gov attribute store)

Maybe the GOV.UK Attribute Service can do those things as well as storing the GOV.UK-specific attributes?

Reply:
> That's a very good point, it is a bit of a weird "frontend" app. Maybe the login / logout functionality should live in frontend and the attribute management part live in a separate API service which is kept properly internal.

I think that makes sense. Having the attribute management service as an internal app will help us keep the API private. Then if an attribute needs to be accessed by a third-party then that should make the case for the attribute to be available in the GOV.UK Account.

> Maybe the GOV.UK Attribute Service can do those things as well as storing the GOV.UK-specific attributes?

Yes, that would make sense. An attribute service for GOV.UK publishing should be responsive enough to support handling sessions.

Reply:
I'll update the RFC to introduce a new internal API app for the endpoints already described, and say that `frontend` will handle the user-facing redirection flows.

Reply:
d80b53187662942a351375f99a86ea2d91b29667

Reply:
> I do wonder whether a broad app like /frontend might still be a potential home since that's the somewhat custom frontend app (@kevindew's first comment)

If we're going to split login and attribute storage, then might an existing frontend app be a good candidate for the `/sign-in*` and `/sign-out` public endpoints?